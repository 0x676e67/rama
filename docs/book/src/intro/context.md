# ☀️ Context\<State\>

To recap, a `Service::serve` method has the following signature:

```rust,noplayground
fn serve(
    &self,
    ctx: Context<State>,
    req: Request,
) -> impl Future<Output = Result<Self::Response, Self::Error>> + Send + '_;
```

- `&self` allows services to access shared `Send + Sync + 'static` state internal and specific to that `Service`;
- `Request` is the input used to produce a `Result`.

`Context<State>` is what this chapter is about,
and its documemtation can be consumed at <https://ramaproxy.org/docs/rama/service/context/struct.Context.html>.

`Context<State>` is used to:

- access shared typesafe `State` defined by the code location instantiating the `Service` on its own or part of a _stack_.
- access `Extensions` that can be used to dynamically get and set extra (optional) data, to be passed for usage by inner service(s).
- spawn tasks for the given (async) executor, doing so gracefully if configured to do so.
- abrubt tasks early in a graceful manner in case of a shutdown using the gracuful `ShutdownGuard` if defined.

This is a clear distinction from a `Tower` service which only takes a `Request`.
If that `Request` is an `http Request` it does allow one to add extra optional data using
the `Extensions` type/data also available in an `http Request`. Hower it provides no means
of typesafe `State`, executors, spawning etc. On top of that it would make it more awkward to
also freely pass all this data between services, especially those operating
across different layers of the network.

## State

rama` supports two kinds of states:

1. type-safe state: this is the `S` generic parameter in [`Context`] and is to be used
   as much as possible, given its existence and type properties can be validated at compile time
2. dynamic state: these can be injected as [`Extensions`]s using methods such as [`Context::insert`]

As a rule of thumb one should use the type-safe state (1) in case:

- the state is always expected to exist at the point the middleware/service is called
- the state is specific to the app or middleware
- and the state can be constructed in a default/empty state

The latter is important given the state is often created (or at least reserved) prior to
it is actually being populated by the relevant middleware. This is not the case for app-specific state
such as Database pools which are created since the start and shared amongs many different tasks.

The rule could be be simplified to "if you need to `.unwrap()` you probably want type-safe state instead".
It's however just a guideline and not a hard rule. As maintainers of `rama` we'll do our best to respect it though,
and we recommend you to do the same.

Any state that is optional, and especially optional state injected by middleware, can be inserted using extensions.
It is however important to try as much as possible to then also consume this state in an approach that deals
gracefully with its absence. Good examples of this are header-related inputs. Headers might be set or not,
and so absence of [`Extensions`]s that might be created as a result of these might reasonably not exist.
It might of course still mean the app returns an error response when it is absent, but it should not unwrap/panic.

[`Context`]: https://ramaproxy.org/docs/rama/service/context/struct.Context.html
[`Context::insert`]: https://ramaproxy.org/docs/rama/service/context/struct.Context.html#method.insert
[`Extensions`]: https://ramaproxy.org/docs/rama/service/context/struct.Extensions.html

## Extensions

Rama makes no use of `http::Extensions`, and instead keeps it all within the extensions API provided by `Context<State>`.
Extensions are there for middleware to inject state into the context. Here is how you should think about extensions:

- Extensions are not for shared data, as it will be unique to this request path, with only the inner service having access to it.
  - Parent of sibling services will not have access to newly inserted data;
- Extensions can be be referenced (`get`) and inserted (`insert`). They cannot be exclusively referenced (`mut`),
  nor can they be removed or anything ese.

`State` in contrast is shared and is fully statically typed. This is a great solution for putting resource pools, databases, etc...
However such a statically typed static type does not lend itself well towards state that is provided by an extension and that is unique to
a specific request (path).

