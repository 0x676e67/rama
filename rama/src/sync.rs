pub use std::sync::{
    atomic::{
        AtomicBool, AtomicI16, AtomicI32, AtomicI64, AtomicI8, AtomicIsize, AtomicPtr, AtomicU16,
        AtomicU32, AtomicU64, AtomicU8, AtomicUsize, Ordering as AtomicOrdering,
    },
    Arc, Mutex,
};

pub use tokio::sync::Mutex as AsyncMutex;
