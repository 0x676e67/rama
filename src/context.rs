//! Context passed to and between services as input.
//!
//! # State
//!
//! [`rama`] supports two kinds of states:
//!
//! 1. type-safe state: this is the `S` generic parameter in [`Context`] and is to be used
//!    as much as possible, given its existence and type properties can be validated at compile time
//! 2. dynamic state: these can be injected as [`Extensions`]s using methods such as [`Context::insert`]
//!
//! As a rule of thumb one should use the type-safe state (1) in case:
//!
//! - the state is always expected to exist at the point the middleware/service is called
//! - the state is specific to the app or middleware
//! - and the state can be constructed in a default/empty state
//!
//! The latter is important given the state is often created (or at least reserved) prior to
//! it is actually being populated by the relevant middleware. This is not the case for app-specific state
//! such as Database pools which are created since the start and shared among many different tasks.
//!
//! The rule could be be simplified to "if you need to `.unwrap()` you probably want type-safe state instead".
//! It's however just a guideline and not a hard rule. As maintainers of [`rama`] we'll do our best to respect it though,
//! and we recommend you to do the same.
//!
//! Any state that is optional, and especially optional state injected by middleware, can be inserted using extensions.
//! It is however important to try as much as possible to then also consume this state in an approach that deals
//! gracefully with its absence. Good examples of this are header-related inputs. Headers might be set or not,
//! and so absence of [`Extensions`]s that might be created as a result of these might reasonably not exist.
//! It might of course still mean the app returns an error response when it is absent, but it should not unwrap/panic.
//!
//! [`rama`]: crate
//!
//! ## State Wraps
//!
//! [`rama`] was built from the ground up to operate on and between different layers of the network stack.
//! This has also an impact on state. Because sure, typed state is nice, but state leakage is not. What do I mean with that?
//!
//! When creating a [`TcpListener`] with state you are essentially creating and injecting state, which will remain
//! as "read-only" for the enire life cycle of that [`TcpListener`] and to be made available for every incoming _tcp_ connection,
//! as well as the application requests (Http requests). This is great for stuff that is okay to share, but it is not desired
//! for state that you wish to have a narrower scope. Examples are state that are tied to a single _tcp_ connection and thus
//! you do not wish to keep a global cache for this, as it would either be shared or get overly complicated to ensure
//! you keep things separate and clean.
//!
//! The solution is to wrap your state.
//!
//! > Example: [http_conn_state.rs](https://github.com/plabayo/rama/tree/main/examples/http_conn_state.rs)
//!
//! The above example shows how can use the [`#as_ref(wrap)`] property within an `#[derive(AsRef)]` derived "state" struct,
//! to wrap in a type-safe manner the "app-global" state within the "conn-specific" (tcp) state. This allows you to have
//! state freshly created for each connection while still having ease of access to the global state.
//!
//! Note though that you do not need the `AsRef` macro or even trait implementation to get this kind of access in your
//! own app-specific leaf services. It is however useful — and at times even a requirement — in case you want your
//! middleware stack to also include generic middleware that expect `AsRef<T>` trait bounds for type-safe access to
//! state from within a middleware. E.g. in case your middleware expects a data source for some specific data type,
//! it is of no use to have that middleware compile without knowing for sure that data source is made available
//! to that middleware.
//!
//! [`TcpListener`]: crate::tcp::server::TcpListener

#[doc(inline)]
pub use ::rama_core::context as context;

#[doc(inline)]
pub use ::rama_macros::AsRef;
