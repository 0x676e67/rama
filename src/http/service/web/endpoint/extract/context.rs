use super::FromRequest;
use crate::http::Request;
use crate::Context;
use std::convert::Infallible;

impl<S> FromRequest<S> for Context<S>
where
    S: Send + Sync + 'static,
{
    type Rejection = Infallible;

    async fn from_request(ctx: Context<S>, _req: Request) -> Result<Self, Self::Rejection> {
        Ok(ctx)
    }
}
