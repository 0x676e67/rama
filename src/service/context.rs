//! Context passed to and between services as input.

/// Context passed to and between services as input.
#[derive(Debug)]
pub struct Context<S> {
    state: S,
}

impl<S> Context<S> {
    /// Create a new [`Context`] with the given state.
    pub fn new(state: S) -> Self {
        Self { state }
    }

    /// Get a reference to the state.
    pub fn state(&self) -> &S {
        &self.state
    }

    /// Get a mutable reference to the state.
    pub fn state_mut(&mut self) -> &mut S {
        &mut self.state
    }
}
