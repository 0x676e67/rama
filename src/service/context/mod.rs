//! Context passed to and between services as input.

use std::{future::Future, sync::Arc};
use tokio::task::JoinHandle;

mod extensions;
pub use extensions::Extensions;
use tokio_graceful::ShutdownGuard;

use crate::rt::Executor;

/// Context passed to and between services as input.
#[derive(Debug)]
pub struct Context<S> {
    state: Arc<S>,
    executor: Executor,
    extensions: Extensions,
}

impl Default for Context<()> {
    fn default() -> Self {
        Self::new(Arc::new(()), Executor::default())
    }
}

impl<S> Clone for Context<S> {
    fn clone(&self) -> Self {
        Self {
            state: self.state.clone(),
            executor: self.executor.clone(),
            extensions: self.extensions.clone(),
        }
    }
}

impl<S> Context<S> {
    /// Create a new [`Context`] with the given state.
    pub fn new(state: Arc<S>, executor: Executor) -> Self {
        Self {
            state,
            executor,
            extensions: Extensions::new(),
        }
    }

    /// Get a reference to the state.
    pub fn state(&self) -> &S {
        &self.state
    }

    /// Get a reference to the executor.
    pub fn executor(&self) -> &Executor {
        &self.executor
    }

    /// Spawn a future on the current executor,
    /// this is spawned gracefully in case a shutdown guard has been registered.
    pub fn spawn<F>(&self, future: F) -> JoinHandle<F::Output>
    where
        F: Future + Send + 'static,
        F::Output: Send + 'static,
    {
        self.executor.spawn_task(future)
    }

    /// Get a reference to an extension.
    ///
    /// An extension is a type that implements `Send + Sync + 'static`,
    /// and can be used to inject dynamic data into the [`Context`].
    ///
    /// Extensions are specific to this [`Context`]. It will be cloned when the [`Context`] is cloned,
    /// but extensions inserted using [`Context::insert`] will not be visible the
    /// original [`Context`], or any other cloned [`Context`].
    ///
    /// Please use the statically typed state (see [`Context::state`]) for data that is shared between
    /// all context clones, parent or not.
    ///
    /// # Example
    ///
    /// ```
    /// # use rama::service::Context;
    /// # let mut ctx = Context::default();
    /// # ctx.insert(5i32);
    /// assert_eq!(ctx.get::<i32>(), Some(&5i32));
    /// ```
    pub fn get<T: Send + Sync + 'static>(&self) -> Option<&T> {
        self.extensions.get::<T>()
    }

    /// Insert an extension into the [`Context`].
    ///
    /// If a extension of this type already existed, it will
    /// be returned.
    ///
    /// See [`Context::get`] for more details regarding extensions.
    ///
    /// # Example
    ///
    /// ```
    /// # use rama::service::Context;
    /// let mut ctx = Context::default();
    ///
    /// assert_eq!(ctx.insert(5i32), None);
    /// assert_eq!(ctx.get::<i32>(), Some(&5i32));
    ///
    /// assert_eq!(ctx.insert(4i32), Some(5i32));
    /// assert_eq!(ctx.get::<i32>(), Some(&4i32));
    /// ```
    pub fn insert<T: Clone + Send + Sync + 'static>(&mut self, extension: T) -> Option<T> {
        self.extensions.insert(extension)
    }

    /// Extend The [`Context`] [`Extensions`] with another [`Extensions`].
    ///
    /// # Example
    ///
    /// ```
    /// # use rama::service::{context::Extensions, Context};
    /// let mut ctx = Context::default();
    /// let mut ext = Extensions::default();
    ///
    /// ctx.insert(true);
    /// ext.insert(5i32);
    /// ctx.extend(ext);
    ///
    /// assert_eq!(ctx.get::<bool>(), Some(&true));
    /// assert_eq!(ctx.get::<i32>(), Some(&5i32));
    /// ```
    pub fn extend(&mut self, extensions: Extensions) {
        self.extensions.extend(extensions);
    }

    /// Clear the [`Context`] of all inserted [`Extensions`].
    ///
    /// # Example
    ///
    /// ```
    /// # use rama::service::Context;
    /// let mut ctx = Context::default();
    ///
    /// ctx.insert(5i32);
    /// assert_eq!(ctx.get::<i32>(), Some(&5i32));
    ///
    /// ctx.clear();
    /// assert_eq!(ctx.get::<i32>(), None);
    /// ```
    pub fn clear(&mut self) {
        self.extensions.clear();
    }

    /// Get a reference to the shutdown guard,
    /// if and only if the context was created within a graceful environment.
    pub fn guard(&self) -> Option<&ShutdownGuard> {
        self.executor.guard()
    }

    /// Turn this Context into a parent [`Context`].
    ///
    /// Naming is hard. Essentially it is meant to optimise the [`Context`] for cloning,
    /// so that the extensions are not cloned upon [`Context`] cloning, but instead
    /// are shared between the original [`Context`] and the cloned [`Context`].
    ///
    /// This is used when branching the [`Context`] into multiple [`Context`]s,
    /// e.g. to go from a Transport Layer to a HTTP Layer, where the context is now
    /// branched for each HTTP request.
    pub fn into_parent(self) -> Self {
        Self {
            state: self.state,
            executor: self.executor,
            extensions: self.extensions.into_parent(),
        }
    }
}
