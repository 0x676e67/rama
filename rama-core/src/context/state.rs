use super::Context;
use std::convert::Infallible;

/// Transforms the `State` embedded in the input [`Context`]
/// into the [`StateTransformer::Output`] state.
///
/// The transformer can return [`StateTransformer::Error`]
/// in case the transformation could not succeed. A common example
/// could be when depending on the availability of state, or
/// validation of dynamic (extension) state.
pub trait StateTransformer<Input> {
    /// Transformed `State` object.
    type Output: Clone + Send + Sync + 'static;
    /// Error that is returned in case the transformation failed.
    type Error;

    /// Transforms the state from the input [`Context`] either in a new state object,
    /// or an error should it have failed.
    fn transform_state(&self, ctx: &Context<Input>) -> Result<Self::Output, Self::Error>;
}

impl<Input> StateTransformer<Input> for ()
where
    Input: Clone + Send + Sync + 'static,
{
    type Output = Input;
    type Error = Infallible;

    fn transform_state(&self, ctx: &Context<Input>) -> Result<Self::Output, Self::Error> {
        Ok(ctx.state().clone())
    }
}

impl<F, Input, Output, Error> StateTransformer<Input> for F
where
    F: Fn(&Context<Input>) -> Result<Output, Error>,
    Input: Clone + Send + Sync + 'static,
    Output: Clone + Send + Sync + 'static,
{
    type Output = Output;
    type Error = Error;

    fn transform_state(&self, ctx: &Context<Input>) -> Result<Self::Output, Self::Error> {
        (self)(ctx)
    }
}

#[cfg(test)]
mod test {
    use rama_error::OpaqueError;

    use crate::context::StateTransformer;
    use crate::rt::Executor;
    use crate::Context;

    #[test]
    fn test_state_transform_default() {
        #[derive(Debug, Clone)]
        struct State {
            answer: usize,
        }

        let state =
            ().transform_state(&Context::new(State { answer: 42 }, Executor::default()))
                .unwrap();
        assert_eq!(state.answer, 42);
    }

    #[test]
    fn test_state_transform_custom() {
        #[derive(Debug, Clone)]
        struct State {
            answer: usize,
        }

        #[derive(Debug, Clone)]
        struct OutputState {
            answer: usize,
            multiplier_used: Multiplier,
        }

        #[derive(Debug, Clone)]
        struct Multiplier(usize);

        fn transformer(ctx: &Context<State>) -> Result<OutputState, OpaqueError> {
            let multiplier = ctx
                .get::<Multiplier>()
                .ok_or_else(|| OpaqueError::from_display("missing multipier"))?;
            Ok(OutputState {
                answer: ctx.state().answer * multiplier.0,
                multiplier_used: multiplier.clone(),
            })
        }

        let mut ctx = Context::new(State { answer: 21 }, Executor::default());

        assert!(transformer.transform_state(&ctx).is_err());

        ctx.insert(Multiplier(2));

        let state = transformer.transform_state(&ctx).unwrap();
        assert_eq!(state.answer, 42);
        assert_eq!(state.multiplier_used.0, 2);
    }
}
