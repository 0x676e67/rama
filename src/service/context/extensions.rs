use std::any::{Any, TypeId};
use std::collections::HashMap;
use std::fmt;
use std::hash::{BuildHasherDefault, Hasher};
use std::sync::Arc;

type AnyMap = HashMap<TypeId, Box<dyn AnyClone + Send + Sync>, BuildHasherDefault<IdHasher>>;

// With TypeIds as keys, there's no need to hash them. They are already hashes
// themselves, coming from the compiler. The IdHasher just holds the u64 of
// the TypeId, and then returns it, instead of doing any bit fiddling.
#[derive(Default)]
struct IdHasher(u64);

impl Hasher for IdHasher {
    fn write(&mut self, _: &[u8]) {
        unreachable!("TypeId calls write_u64");
    }

    #[inline]
    fn write_u64(&mut self, id: u64) {
        self.0 = id;
    }

    #[inline]
    fn finish(&self) -> u64 {
        self.0
    }
}

/// A type map of protocol extensions.
///
/// `Extensions` can be used by `Request` and `Response` to store
/// extra data derived from the underlying protocol.
#[derive(Clone, Default)]
pub struct Extensions {
    // If extensions are never used, no need to carry around an empty HashMap.
    // That's 3 words. Instead, this is only 1 word.
    map: Option<Box<AnyMap>>,
}

impl Extensions {
    /// Create an empty `Extensions`.
    #[inline]
    pub fn new() -> Extensions {
        Extensions { map: None }
    }

    /// Insert a type into this `Extensions`.
    ///
    /// If a extension of this type already existed, it will
    /// be returned.
    pub fn insert<T: Clone + Send + Sync + 'static>(&mut self, val: T) -> Option<T> {
        self.map
            .get_or_insert_with(Box::default)
            .insert(TypeId::of::<T>(), Box::new(val))
            .and_then(|boxed| boxed.into_any().downcast().ok().map(|boxed| *boxed))
    }

    /// Extend these extensions with another Extensions.
    pub fn extend(&mut self, other: Extensions) {
        if let Some(other_map) = other.map {
            let map = self.map.get_or_insert_with(Box::default);
            #[allow(clippy::useless_conversion)]
            map.extend(other_map.into_iter());
        }
    }

    /// Clear the `Extensions` of all inserted extensions.
    pub fn clear(&mut self) {
        if let Some(map) = self.map.as_mut() {
            map.clear();
        }
    }
    /// Get an extension or T's [default()](std::default::Default).
    /// Refer to [Context::get](Self::get) for more detail's
    pub fn get_or_insert_default<T: Send + Default + Clone + Sync + 'static>(&mut self) -> &mut T {
        self.get_or_insert_with(|| T::default())
    }
    /// Retrieves a value of type `T` from the context. If the value does not exist,
    /// returns the provided value.
    /// Refer to [Context::get](Self::get) for more detail's
    pub fn get_or_insert<T: Send + Clone + Sync + 'static>(&mut self, fallback: T) -> &T {
        let b = match &self.map {
            Some(boxed_map) => boxed_map.contains_key(&TypeId::of::<T>()),
            None => false,
        };
        if !b {
            self.insert(fallback);
        }
        // At this point, it is safe to unwrap because we know the value exists
        self.get::<T>().unwrap()
    }
    /// Inserts a value into the map computed from `f` into if it is [`None`],
    /// then returns a mutable reference to the contained value.
    pub fn get_or_insert_with<T: 'static + Send + Sync + Clone>(
        &mut self,
        f: impl FnOnce() -> T,
    ) -> &mut T {
        let type_id = TypeId::of::<T>();

        let map = self.map.get_or_insert_with(Default::default);

        let boxed_any = map.entry(type_id).or_insert_with(|| Box::new(f()));

        let any_ref_mut = boxed_any.as_any_mut();
        any_ref_mut.downcast_mut::<T>().expect("Type mismatch")
    }

    /// Get a reference to a type previously inserted on this `Extensions`.
    pub fn get<T: Send + Sync + 'static>(&self) -> Option<&T> {
        self.get_inner::<T>().or_else(|| {
            self.get_inner::<ParentExtensions>()
                .and_then(|parent| parent.extensions.get::<T>())
        })
    }

    fn get_inner<T: Send + Sync + 'static>(&self) -> Option<&T> {
        self.map
            .as_ref()
            .and_then(|map| map.get(&TypeId::of::<T>()))
            .and_then(|boxed| (**boxed).as_any().downcast_ref())
    }

    /// Get a new extension map with the current extensions as parent.
    ///
    /// Note that later edits to parent won't be reflected here.
    pub fn into_parent(self) -> Extensions {
        let mut ext = Extensions::new();
        ext.insert(ParentExtensions {
            extensions: Arc::new(self),
        });
        ext
    }
    /// Checks if a value of type T exists in the extensions map
    pub fn contains<T: 'static + Send + Sync>(&self) -> bool {
        match &self.map {
            Some(boxed_map) => boxed_map.contains_key(&TypeId::of::<T>()),
            None => false,
        }
    }
}

#[derive(Debug, Clone)]
struct ParentExtensions {
    extensions: Arc<Extensions>,
}

impl fmt::Debug for Extensions {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("Extensions").finish()
    }
}

trait AnyClone: Any {
    fn clone_box(&self) -> Box<dyn AnyClone + Send + Sync>;
    fn as_any(&self) -> &dyn Any;
    fn as_any_mut(&mut self) -> &mut dyn Any;
    fn into_any(self: Box<Self>) -> Box<dyn Any>;
}

impl<T: Clone + Send + Sync + 'static> AnyClone for T {
    fn clone_box(&self) -> Box<dyn AnyClone + Send + Sync> {
        Box::new(self.clone())
    }

    fn as_any(&self) -> &dyn Any {
        self
    }

    fn as_any_mut(&mut self) -> &mut dyn Any {
        self
    }

    fn into_any(self: Box<Self>) -> Box<dyn Any> {
        self
    }
}

impl Clone for Box<dyn AnyClone + Send + Sync> {
    fn clone(&self) -> Self {
        (**self).clone_box()
    }
}

#[test]
fn test_extensions() {
    #[derive(Clone, Debug, PartialEq)]
    struct MyType(i32);

    let mut extensions = Extensions::new();

    extensions.insert(5i32);
    extensions.insert(MyType(10));

    assert_eq!(extensions.get(), Some(&5i32));

    let mut ext2 = extensions.clone();
    let mut ext3 = extensions.into_parent();

    ext2.insert(true);
    ext3.insert(false);

    assert_eq!(ext2.get(), Some(&5i32));
    assert_eq!(ext2.get(), Some(&MyType(10)));
    assert_eq!(ext2.get(), Some(&true));

    assert_eq!(ext3.get(), Some(&5i32));
    assert_eq!(ext3.get(), Some(&MyType(10)));
    assert_eq!(ext3.get(), Some(&false));

    // test extend
    let mut extensions = Extensions::new();
    extensions.insert(5i32);
    extensions.insert(MyType(10));

    let mut extensions2 = Extensions::new();
    extensions2.extend(extensions);
    assert_eq!(extensions2.get(), Some(&5i32));
    assert_eq!(extensions2.get(), Some(&MyType(10)));

    // test clear
    extensions2.clear();
    assert_eq!(extensions2.get::<i32>(), None);
    assert_eq!(extensions2.get::<MyType>(), None);
}
