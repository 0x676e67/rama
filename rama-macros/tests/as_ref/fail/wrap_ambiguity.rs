use rama::context::AsRef;
use std::sync::Arc;

#[derive(Clone, AsRef)]
struct AppState {
    auth_token: String,
    #[as_ref(skip)]
    also_string: String,
}

#[derive(AsRef)]
struct ConnState {
    #[as_ref(wrap)]
    app: AppState,
    auth_token: String,
}

fn main() {}
