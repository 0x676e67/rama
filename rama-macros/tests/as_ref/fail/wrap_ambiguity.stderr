error[E0119]: conflicting implementations of trait `AsRef<std::string::String>` for type `ConnState`
  --> tests/as_ref/fail/wrap_ambiguity.rs:15:17
   |
14 |     app: Arc<AppState>,
   |          --- first implementation here
15 |     auth_token: String,
   |                 ^^^^^^ conflicting implementation for `ConnState`
